alf/algorithms/muzero_representation_learner.py

```diff
@@ -930,3 +931,160 @@ def __call__(self, age, max_td_steps, current_max_age):
+            reward_spec=reward_spec,
+            train_state_spec=(),
+            config=None,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        self._training_options = training_options
+
+        # Override the training behavior related parameters in the config when
+        # ``training_options`` is explicitly provided. This is done before
+        # calling the ``__init__`` of the super class.
+        updated = copy.copy(config)
+        updated.whole_replay_buffer_training = False
+        updated.clear_replay_buffer = False
+        updated.mini_batch_length = training_options.mini_batch_length
+        updated.mini_batch_size = training_options.mini_batch_size
+        updated.num_updates_per_train_iter = training_options.num_updates_per_train_iter
+        updated.replay_buffer_length = training_options.replay_buffer_length
+        updated.initial_collect_steps = training_options.initial_collect_steps
+        updated.priority_replay = training_options.priority_replay
+        updated.priority_replay_alpha = as_scheduler(
+            training_options.priority_replay_alpha)
+        updated.priority_replay_beta = as_scheduler(
+            training_options.priority_replay_beta)
+
+        self._impl = impl_cls(
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            reward_spec=reward_spec,
+            config=updated,
+            debug_summaries=debug_summaries)
+        self._impl.force_params_visible_to_parent = True
+        assert self._impl._reanalyze_ratio in [
+            0.0, 1.0
+        ], ('Currently MuzeroRepresentationLearner only support reanalyze ratio 0.0 or 1.0'
+            )
+
+        if self._impl._reanalyze_ratio > 0:
+            assert config.use_rollout_state, (
+                'use_rollout_state needs to be True when reanalyze is used.')
+
+    @property
+    def output_spec(self):
+        """Access the spec of the produced representation.
+
+        This will be used as the obervation spec for the subsequent RL
+        algorithm.
+
+        """
+        return self._impl._model.repr_spec
+
+    def rollout_step(self, time_step: TimeStep, state):
+        repr_step = self._impl.rollout_step(time_step, state)
+        # Save in the representation learner's own replay buffer. Note that
+        # ``observe_for_replay`` when called for the first time will have the
+        # side effect of creating the replay buffer.
+        if self._impl._replay_buffer is None:
+            self._impl.set_replay_buffer(
+                time_step.env_id.shape[0],
+                self._training_options.replay_buffer_length,
+                self._training_options.priority_replay)
+        exp = make_experience(time_step.untransformed, repr_step, state)
+        self._impl.observe_for_replay(exp)
+        return repr_step
+
+    def train_step(self, exp: TimeStep, state, rollout_info):
+        return self._impl.predict_step(exp, state)
```

Comment:
`impl.rollout_step`?

Reply:
Updated to `rollout_step`. I did not quite understand the difference though (they are the same except for the `enable_amp` part). Can you share some thoughts? Thanks.